pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,277 @@
+		signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey] = now.Add(
+			time.Duration(s.config.GetVerificationCodeExpiresInMin()) * time.Minute).Format(TimestampLayout)
+
+		// Generate the verification message with the new verification code
+		content := fmt.Sprintf(s.config.GetVerificationMessageTemplate(), verificationCode)
+		client := twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), s.HttpClient)
+		msg, err := client.Messages.SendMessage(s.config.GetTwilioFromNumber(), e164PhoneNumber, content, nil)
+		if err != nil {
+			if msg != nil {
+				log.Error(ctx, err, fmt.Sprintf("error while sending, code: %d message: %s", msg.ErrorCode, msg.ErrorMessage))
+			} else {
+				log.Error(ctx, err, "unknown error while sending")
+			}
+
+			// If we get an error here then just die, don't bother updating the UserSignup
+			return errors.NewInternalError(err, "error while sending verification code")
+		}
+	}
+
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, err, "error while updating UserSignup resource")
+		return errors.NewInternalError(updateErr, "error while updating UserSignup resource")
+	}
+
+	return initError
+}
+
+func generateVerificationCode() string {
+	buf := make([]byte, codeLength)
+	rand.Read(buf)
+
+	charsetLen := len(codeCharset)
+	for i := 0; i < codeLength; i++ {
+		buf[i] = codeCharset[int(buf[i])%charsetLen]
+	}
+
+	return string(buf)
+}
+
+// VerifyCode validates the user's phone verification code.  It updates the specified UserSignup value, so even
+// if an error is returned by this function the caller should still process changes to it
+func (s *ServiceImpl) VerifyCode(ctx *gin.Context, userID string, code string) (verificationErr error) {
+
+	// If we can't even find the UserSignup, then die here
+	signup, lookupErr := s.Services().SignupService().GetUserSignup(userID)
+	if lookupErr != nil {
+		if apierrors.IsNotFound(lookupErr) {
+			log.Error(ctx, lookupErr, "usersignup not found")
+			return errors.NewNotFoundError(lookupErr, "user not found")
+		}
+		log.Error(ctx, lookupErr, "error retrieving usersignup")
+		return errors.NewInternalError(lookupErr, fmt.Sprintf("error retrieving usersignup: %s", userID))
+	}
+
+	now := time.Now()
+
+	// If 24 hours has passed since the verification timestamp, then reset the timestamp and verification attempts
+	ts, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey])
+	if parseErr != nil || (now.After(ts.Add(24 * time.Hour))) {
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = "0"
+	}
+
+	attemptsMade, convErr := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
```

Comment:
I maybe got lost in the usage of the annotations. 
In the method `InitVerification` that initiates the verification phase and sets/sends the expected verification code, there you use:
* UserSignupVerificationInitTimestampAnnotationKey
* UserSignupVerificationCounterAnnotationKey

The purpose is clear - we don't want to let users to generate the code too many times in a day.

In this method `VerifyCode` that expects the generated code from the previous step, there you use:
* UserSignupVerificationTimestampAnnotationKey
* UserVerificationAttemptsAnnotationKey

I guess that the purpose is allowing the user to verify the same code only a few times and not infinitely, right? If yes then I have these questions:
1. why don't we reset the counter `UserVerificationAttemptsAnnotationKey` when we generate a new verification code?
2. why do we check the 24 hours timeout when the code has validity only a few minutes?
3. why do we have the `UserSignupVerificationTimestampAnnotationKey` at all? The time limitation should be done by the code expiry, shouldn't it?

Reply:
We limit the total number of attempts allowed per day, regardless of how many times the user has requested a verification code.  The `UserSignupVerificationTimestampAnnotationKey` annotation is used in conjunction with `UserVerificationAttemptsAnnotationKey` in order to track the number of verification attempts a user has made within a 24 hour period.  Once they exceed the maximum allowed, they can no longer attempt to verify until the 24 hour period has lapsed.

Reply:
ok, but wouldn't we achieve the same result if we allowed for example one attempt per verification code?

Apart from that - looking at the default values in the config:
`DefaultVerificationDailyLimit = 5`
`DefaultVerificationAttemptsAllowed = 3`
which means that I can request the verification code 5 times in a day, but can try to verify the code only 3 times...? Imagine this situation:
1. I request a verification code
2. I try to verify the code 3 times but for some reason it fails every time
3. I can still request a new code - actually four new codes - but cannot verify any of them

That doesn't make any sense :man_shrugging:

Reply:
I think these values should probably be swapped around, that would seem to make more sense then.

Reply:
but still, it will be the same UX - you can do multiple verification of the same code - even if we swap the values - still you could run out of the attempts for the first one or two codes for the day, then no attempt will be left for the last generated code.

We should either use the same annotation/constant for both cases (generation and attempts) or limit the number of attempts per code, not per day.

Reply:
Yeah.. I think it's reasonable to either:
- Count the overall number of verification code generations/sms per day.
- Count the number of verification attempts per code (not overall).
Or:
- Simplify the logic and use one parameter for both the code generation and overall attempts per day.

The first one is preferable but I think it's fine to keep it simple if it really simplify the code.

Reply:
Guys, let's deal with this in a followup issue as we've gotten off track here.  This PR is about the architectural refactoring only, and while I agree that we should look more closely at the verification workflow business logic it is out of scope of this PR.

Reply:
Sure. Let's address this in a separate PR.

Reply:
ok, I'm fine with doing it in separate PR. 
@sbryzak Just please create jira issue for that, so we won't forget, and we can plan it. Thanks

Reply:
https://issues.redhat.com/browse/CRT-839